signals/signals.go

```diff
@@ -42,21 +40,23 @@ func handleSignals(finalizer func(), finishedChan chan struct{}) {
 	signal.Notify(sigChan, syscall.SIGTERM, syscall.SIGINT, syscall.SIGKILL)
 	defer signal.Stop(sigChan)
 	if finalizer != nil {
-		select {
-		case sig := <-sigChan:
-			// first signal use finalizer to terminate cleanly
+		switch sig := <-sigChan; sig {
+		// SIGINT & SIGTERM: use finalizer & terminate cleanly
+		case syscall.SIGINT, syscall.SIGTERM:
 			log.Logvf(log.Always, "signal '%s' received; attempting to shut down", sig)
 			finalizer()
-		case <-finishedChan:
-			return
+			os.Exit(util.ExitFailure)
+		// SIGKILL: terminate immediately
+		case syscall.SIGKILL:
+			log.Logvf(log.Always, "signal '%s' received; forcefully terminating", sig)
+			os.Exit(util.ExitFailure)
 		}
 	}
+
 	select {
 	case sig := <-sigChan:
-		// second signal exits immediately
+		// exit immediately
 		log.Logvf(log.Always, "signal '%s' received; forcefully terminating", sig)
 		os.Exit(util.ExitFailure)
```

Comment:
Since there is only one channel to wait on, we no longer need a `select` statement here. I believe we can update this to
```
sig := <-sigChan
log.Logvf(...)
os.Exit(...)
```
since the first line there (reading from the channel) will block until a signal comes in.

Reply:
Good catch! I added the changes to an else statement to make it clearer that this is the expected behavior if a finalizer is not defined.